api/src/auth_routes.ts

```diff
@@ -63,9 +76,57 @@ export function determine_callback_urls(provider_name: string): {
   };
 }
 
-export function add_auth_routes(app: any, handlers: any) {
-  app.get('/auth/', (req: any, res: any) => {
-    // Allow the user to decide what auth mechanism to use
+/**
+ * Check that a redirect URL is one that we allow
+ * Mainly want to avoid a non-url redirect. Without a whitlist of
+ * app URLs we can't really block any but check that it starts with 'http'
+ * For mobile we redirect to the custom URL scheme for the app
+ * currently hard-coded as `org.fedarch.faims3` but should be configurable
+ *
+ * @param redirect URL to redirect to
+ * @returns a valid URl to redirect to, default to '/' if
+ *   the one passed in is bad
+ */
+function validateRedirect(redirect: string) {
+  if (redirect.startsWith('http')) {
+    // could match against a whitelist of allowed URLs but for now...
+    return redirect;
+  } else if (
+    redirect.startsWith('/') ||
+    redirect.startsWith('org.fedarch.faims3://')
+  ) {
+    return redirect;
+  } else {
+    return '/';
+  }
+}
+
+/**
+ * Add authentication routes for local and federated login
+ * The list of handlers are the ids of the configured federated handlers (eg. ['google'])
+ * routes will be set up for each of these for auth and registration
+ * See `auth_providers/index.ts` for registration of providers.
+ *
+ * @param app Express router
+ * @param handlers an array of login provider identifiers
+ */
+export function add_auth_routes(app: Router, handlers: string[]) {
+  // In the following generic type signatures for routes the
+  // order of types is:
+  //  <Params, ResBody, ReqBody, ReqQuery, Locals>
+  // Params: Route parameters
+  // ResBody: Response body
+  // ReqBody: Request body
+  // ReqQuery: Query string parameters
+  // Locals: Response local variables
+
+  app.get<{}, {}, {}, RequestQueryRedirect>('/auth/', async (req, res) => {
```

Comment:
Once we merge new survey work, will be a task to come back and align the validation and typing approach - but appreciate adding the doc strings and interfaces. 👍